relations/src/r1cs/mod.rs

```diff
@@ -1,9 +1,125 @@
-//! Core interface for working with Rank-1 Constraint Systems (R1CS).
+//! Core interfaces for working with Rank-1 Constraint Systems (R1CS), which is
+//! an indexed NP relation.
 
-use ark_std::vec::Vec;
+use crate::{NPRelation, Relation};
+use ark_std::{cmp::Ordering, marker::PhantomData, vec::Vec};
 
-/// A result type specialized to `SynthesisError`.
-pub type Result<T> = core::result::Result<T, SynthesisError>;
+/// An *indexed NP relation* is a relation with an efficient membership check.
+pub struct R1CS<F: Field> {
+    f: PhantomData<F>,
+}
+
+/// An R1CS index consists of three matrices, as well as the number of instance variables
+/// and number of witness variables.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ConstraintMatrices<F: Field> {
+    /// The number of variables that are "public instances" to the constraint
+    /// system.
+    pub num_instance_variables: usize,
+    /// The number of variables that are "private witnesses" to the constraint
+    /// system.
+    pub num_witness_variables: usize,
+    /// The number of constraints in the constraint system.
+    pub num_constraints: usize,
+    /// The number of non_zero entries in the A matrix.
+    pub a_num_non_zero: usize,
+    /// The number of non_zero entries in the B matrix.
+    pub b_num_non_zero: usize,
+    /// The number of non_zero entries in the C matrix.
+    pub c_num_non_zero: usize,
+
+    /// The A constraint matrix.
+    /// `self.mode == SynthesisMode::Prove { construct_matrices = false }`.
+    pub a: Matrix<F>,
+    /// The B constraint matrix.
+    pub b: Matrix<F>,
+    /// The C constraint matrix.
+    pub c: Matrix<F>,
+}
+
+/// An R1CS instance consists of variable assignments to the instance variables.
+/// The first variable must be assigned a value of `F::one()`.
```

Comment:
I wonder if we should do the converse, and make the second variable onwards what goes into the struct?

Reply:
what do other sources of R1CS do, like zkinterface?